### DISCUSSION TITLE
Add `whereWeekday` and `whereWeekdayIn` to the Query Builder

### DISCUSSION THREAD

**Discussion by Arielt99** [laravel/framework#48611] [Ideas]
Hi,

for a date based where, the framework seem to only support the types: year, month, day and time, while as far as I know most of the database management system also support the weekday.
I personnaly often have to create a macro for `whereWeekday` and `whereWeekdayIn`.

I'm prepared to do it, but I'd rather gather some opinions beforehand.

**Comment by shaedrich** [comment]
Would you mind sharing your macros?

**Reply by Arielt99** [reply]
Sure thing, it's made with mysql in mind, if you use another database management system that don't support weekday() you can adapt it.

```
Builder::macro('whereWeekday', function (string $attribute, int $day) {
    /** @var Builder $this  */
    $attribute = implode('.', array_filter(array($this->getModel()?->getTable(), $attribute)));

    $this->whereRaw('weekday('.$attribute.') = '.$day);

    return $this;
});
```

```
Builder::macro('whereWeekdayIn', function (string $attribute, array $days) {
    /** @var Builder $this  */
    $attribute = implode('.', array_filter(array($this->getModel()?->getTable(), $attribute)));

    $this->whereRaw('weekday('.$attribute.') in ('.implode(',', $days).')');

    return $this;
});
```

**Reply by shaedrich** [reply]
Thanks for sharing these!

Can't you just do `$attribute = $this->qualifyColumn($attribute);`?

**Reply by Arielt99** [reply]
Yes you can, it's way better actually ! I didn't know this one, thanks!

**Reply by shaedrich** [reply]
You're welcome!

**Reply by shaedrich** [reply]
> I'm prepared to do it, but I'd rather gather some opinions beforehand. [&hellip;] Sure thing, it's made with mysql in mind, if you use another database management system that don't support weekday() you can adapt it.

If you are going to submit a PR on that matter, good luck with that (you gonna need it)&mdash;this confidence is admirable. Even though, it's common practise (and potentially dangerous in my opinion (not that Taylor Otwell would care)) in Laravel to silently discard database actions that aren't supported by the engine in question, I expect the response to be that they don't want to implement something that isn't supported by each and every database engine.

**Reply by Arielt99** [reply]
If I were to do it, it would be a cleaner way, within the different grammar classes like it's done for the other date based query.

**Reply by shaedrich** [reply]
You could use bindings here:
```diff
-    $this->whereRaw('weekday('.$attribute.') = '.$day);
+    $this->whereRaw('weekday('.$attribute.') = ?', [$day]);
```

It would be nice if it would also support other types than `int` like integer-backed enums

**Comment by henzeb** [comment]
it's best if you look at packages. something like https://github.com/omarelnaghy/lara-date-filter or https://github.com/laracraft-tech/laravel-date-scopes

**Reply by Arielt99** [reply]
I already have my macros for it. My goal with this topic is to implement it in the framework so everyone can use it without having the need to add a dependency.

**Reply by henzeb** [reply]
But still, you need to implement it for every database. It probably means to add methods to the grammar classes... Lots of tests.

**Reply by Arielt99** [reply]
Since it's just adding a supported type to the existing functions, it don't seem that big, but I might have missed some stuff that would make it more difficult.

**Comment by parallels999** [comment]
> I'm prepared to do it, but I'd rather gather some opinions beforehand.

Taylor will close it faster than it will take you to do it. just use the macros, without any review

**Reply by shaedrich** [reply]
One could say, it's his signature move

**Reply by parallels999** [reply]
Always 
>Thanks for your pull request to Laravel!
>
>Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include.
>
>If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions!
>
>If you feel absolutely certain that this code corrects a bug in the framework, please "@" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Reply by shaedrich** [reply]
Well, after all, it's ["A Community Framework"](https://laravel.com/docs/10.x#why-laravel) in name only
